sympy/solvers/tests/test_solveset.py

```diff
@@ -1676,10 +1687,43 @@ def test__is_finite_with_finite_vars():
         Dummy(), Dummy(real=True), Dummy(complex=True)))
     assert f(1/Dummy(real=False)) is True  # b/c it's finite but not 0
 
+
 def test_issue_13550():
     assert solveset(x**2 - 2*x - 15, symbol = x, domain = Interval(-oo, 0)) == FiniteSet(-3)
 
 
 def test_issue_13849():
     t = symbols('t')
     assert nonlinsolve((t*(sqrt(5) + sqrt(2)) - sqrt(2), t), t) == EmptySet()
+
+
+def test_issue_14223():
+    x = Symbol('x')
+    assert solveset((Abs(x + Min(x, 2)) - 2).rewrite(Piecewise), x,
+        S.Reals) == FiniteSet(-1, 1)
+    assert solveset((Abs(x + Min(x, 2)) - 2).rewrite(Piecewise), x,
+        Interval(0, 2)) == FiniteSet(1)
+
+
+def test_issue_10158():
+    x = Symbol('x')
+    dom = S.Reals
+    assert solveset(x*Max(x, 15) - 10, x, dom) == FiniteSet(2/S(3))
+    assert solveset(x*Min(x, 15) - 10, x, dom) == FiniteSet(-sqrt(10), sqrt(10))
+    assert solveset(Max(Abs(x - 3) - 1, x + 2) - 3, x, dom) == FiniteSet(-1, 1)
+    assert solveset(Abs(x - 1) - Abs(y), x, dom) == FiniteSet(-Abs(y) + 1, Abs(y) + 1)
+    assert solveset(Abs(x + 4*Abs(x + 1)), x, dom) == FiniteSet(-4/S(3), -4/S(5))
+    dom = S.Complexes
+    assert solveset(x*Max(x, 15) - 10, x, dom) == FiniteSet(2/S(3))
+    assert solveset(x*Min(x, 15) - 10, x, dom) == FiniteSet(
+        -sqrt(10), sqrt(10))
+    # the following gets rewritten to
```

Comment:
>  do you have any ideas about this situation?

What I figured out @smichr is that this equation gets converted to `Piecewise((x - 1, x + 2 >= Abs(x - 3) - 1), (Abs(x - 3) - 4, True))`, which further divides into `[(x - 1, Interval(0, oo)), (Abs(x - 3) - 4, Interval.open(-oo, 0))]`, now since both these `expr` are in the real domain, for the case of `Abs` _invert_real is called giving the result as {-1}. There is no `ValueError` because `_invert_real` solves it, while in cases for bare Abs in complex domain, there is no implementation in `_invert_complex` therfore they will always go to `_solve_abs` and raise error.

Reply:
> Also Max(x, y) is rewritten assuming that the args are real even though it is not defined for complex values. I'm not sure if that is consistent with Abs(x) not being rewritten unless x is real (but Abs is defined for complex values).

Yes, max or min is rewritten irrespective the args are real or complex [see here](https://github.com/sympy/sympy/blob/master/sympy/functions/elementary/miscellaneous.py#L750), but in case of `Abs` it checks for args to be real, then only rewrites. [see here](https://github.com/sympy/sympy/blob/master/sympy/functions/elementary/complexes.py#L586)

Two things need to be done:

- Restrict rewriting of Min/Max only to real,

- Also, Abs should be made to be able to rewrite in complex domain. Atleast solveset should solve Abs in complex domain (should be solved in another PR)

Reply:
It could be just as simple as only solving Piecewise if the conditions don't involve relationals when the variable is not real:
```
>>> solveset(x<1,x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\leslie\sympy\sympy\solvers\solveset.py", line 1006, in solveset

    setting domain=S.Reals'''))
NotImplementedError:
Inequalities in the complex domain are not supported. Try the real
domain by setting domain=S.Reals
>>> solveset(Eq(x,1),x)
{1}
>>> solveset(Eq(x,I),x)
{I}
```

Reply:
> It could be just as simple as only solving Piecewise if the conditions don't involve relationals when the variable is not real:

Rather than raising error for Piecewise let `_solve_abs` handle it. Committed in your repo.

Reply:
> If you agree with the above I would like to make the changes

Ok, I will raise an error if an attempt to solve a Piecewise with non-Equality conditions in the complex domain is made. Then this can be committed and your changes can be a separate PR. How does that sound?

Reply:
Sounds great!